File: ch4数据结构/queue/queue.go
// package queue is the queue example made up by slice

package queue

// Queue is the common type for queue.
type Queue []interface{}

// Enqueue push element to queue
func (q *Queue) Enqueue(x ...interface{}) {
	for _, val := range x {
		*q = append(*q, val)
	}
}

// Dequeue pop element for queue
func (q *Queue) Dequeue() interface{} {
	queue := *q
	head := queue[0]
	*q = queue[1:]
	return head
}

// IsEmpty return if the queue is empty
func (q *Queue) IsEmpty() bool {
	queue := *q
	return len(queue) == 0
}

File: ch4数据结构/STL/queue.cpp
#include <queue>
#include <iostream>
using namespace std;
int main()
{
    queue<int> q; // 声明一个装 int 类型数据的队列
    q.push(1);    // 入队
    q.push(2);
    q.push(3);
    cout << q.size() << endl; // 输出队列元素个数
    while (!q.empty())
    {                              // 判断队列是否为空
        cout << q.front() << endl; // 访问队首元素
        q.pop();                   // 出队
    }
    return 0;
}